File: ruby/lib/nit/repo.rb
# frozen_string_literal: true

module Nit
  module Repo
    MAX_REPO_NAME_WIDTH = 24

    # Find all git repositories in the given directory (depth 1).
    # Returns a sorted list of absolute paths to directories containing a .git folder.
    def self.find_git_repos(path = Dir.pwd)
      repos = []

      Dir.each_child(path) do |entry|
        full_path = File.join(path, entry)
        next unless File.directory?(full_path)

        git_dir = File.join(full_path, ".git")
        repos << full_path if File.exist?(git_dir)
      end

      repos.sort
    end

    # Extract just the repository name from a path
    def self.repo_name(path)
      File.basename(path) || "unknown"
    end

    # Format repo name with fixed width: truncate long names, pad short ones
    # Returns string like "[name                    ]" (26 chars total)
    def self.format_repo_name(name)
      display_name = if name.length > MAX_REPO_NAME_WIDTH
                       "#{name[0, MAX_REPO_NAME_WIDTH - 4]}-..."
                     else
                       name
                     end
      "[#{display_name.ljust(MAX_REPO_NAME_WIDTH)}]"
    end
  end
end

File: fit-zig/src/repo.zig
const std = @import("std");
const fs = std.fs;

/// Check if the current working directory is inside a git repository.
/// Uses `git rev-parse --git-dir` which correctly handles worktrees,
/// bare repos, and the GIT_DIR environment variable.
pub fn isInsideGitRepo(allocator: std.mem.Allocator) bool {
    const argv = [_][]const u8{ "git", "rev-parse", "--git-dir" };
    var child = std.process.Child.init(&argv, allocator);
    child.stdout_behavior = .Ignore;
    child.stderr_behavior = .Ignore;

    child.spawn() catch return false;
    const term = child.wait() catch return false;
    return term.Exited == 0;
}

/// Find all git repositories in the current directory (depth 1).
/// Returns a sorted list of absolute paths to directories containing a .git folder.
/// Caller owns the returned slice and paths.
pub fn findGitRepos(allocator: std.mem.Allocator) ![][]const u8 {
    const cwd = try std.fs.cwd().realpathAlloc(allocator, ".");
    defer allocator.free(cwd);

    var repos: std.ArrayList([]const u8) = .empty;
    errdefer {
        for (repos.items) |path| {
            allocator.free(path);
        }
        repos.deinit(allocator);
    }

    var dir = try std.fs.cwd().openDir(".", .{ .iterate = true });
    defer dir.close();

    var iter = dir.iterate();
    while (try iter.next()) |entry| {
        if (entry.kind != .directory) continue;

        // Check if .git exists in this directory
        var subdir = dir.openDir(entry.name, .{}) catch continue;
        defer subdir.close();

        subdir.access(".git", .{}) catch continue;

        // Found a git repo - construct full path
        const full_path = try std.fs.path.join(allocator, &.{ cwd, entry.name });
        try repos.append(allocator, full_path);
    }

    // Sort the repos by path
    const items = try repos.toOwnedSlice(allocator);
    std.mem.sort([]const u8, items, {}, struct {
        fn lessThan(_: void, a: []const u8, b: []const u8) bool {
            return std.mem.order(u8, a, b) == .lt;
        }
    }.lessThan);

    return items;
}

/// Extract just the repository name from a path
pub fn repoName(path: []const u8) []const u8 {
    return std.fs.path.basename(path);
}

test "repoName extracts directory name" {
    const name = repoName("/home/user/src/my-repo");
    try std.testing.expectEqualStrings("my-repo", name);
}

test "repoName handles simple name" {
    const name = repoName("my-repo");
    try std.testing.expectEqualStrings("my-repo", name);
}
